### ISSUE TITLE
如何实现字符级别的注意力机制

### ISSUE THREAD

**Issue by huwf9918** [thunlp/OpenNRE#284]
您好，我想实现一个字符级别的注意力机制，于是我就把PCNN+ATT模型中bag级别的注意力机制改了一下，就是单纯的把长度max_len当作bag_size处理，也就相当于把原来 对于一个包内的句子计算的注意力 改成了对于一个句子内的词计算注意力，但效果却很不好，请问能您能帮我解答一下是哪里除了问题吗？

其中，encoder部分的代码如下：
`    def forward(self, label, token, att_mask):
        hidden, CLS = self.bert(token, attention_mask=att_mask)
        hidden=hidden.view(-1,hidden.shape[-1])  # (B, 2H)
        return hidden
`

model部分代码如下：
`    def forward(self, label, token, att_mask, train=True):

        bag_size=self.sentence_encoder.max_length
        rep=self.sentence_encoder(label, token, att_mask)

        # Attention
        if train:

            batch_size = label.size(0)
            query = label.unsqueeze(1)  # (B, 1)
            att_mat = self.fc.weight.data[query]  # (B, 1, H)
            rep = rep.view(batch_size, bag_size, -1)
            att_score = (rep * att_mat).sum(-1)  # (B, bag)
            softmax_att_score = self.softmax(att_score)  # (B, bag)
            bag_rep = (softmax_att_score.unsqueeze(-1) * rep).sum(
                1)  # (B, bag, 1) * (B, bag, H) -> (B, bag, H) -> (B, H)
            bag_rep = self.drop(bag_rep)
            bag_logits = self.fc(bag_rep)  # (B, N)
        else:

            batch_size = rep.size(0) // bag_size
            att_score = torch.matmul(rep, self.fc.weight.data.transpose(0, 1))  # (nsum, H) * (H, N) -> (nsum, N)
            att_score = att_score.view(batch_size, bag_size, -1)  # (B, bag, N)
            rep = rep.view(batch_size, bag_size, -1)  # (B, bag, H)
            softmax_att_score = self.softmax(att_score.transpose(1, 2))  # (B, N, (softmax)bag)
            rep_for_each_rel = torch.matmul(softmax_att_score, rep)  # (B, N, bag) * (B, bag, H) -> (B, N, H)
            bag_logits = self.softmax(self.fc(rep_for_each_rel)).diagonal(dim1=1, dim2=2)  # (B, (each rel)N)

        return bag_logits
`

就是单纯的取了原来bag_attention.py文件中bag_size大于零的部分，并用句子长度max_length作为bag_size处理， 剩下的都没有改动，但是效果却非常不好，请问您能解答一下原因吗，谢谢！
